### PR TITLE
fix(core):fix missing info of flattened_errors

### PR DESCRIPTION
**Pull request by lhanjian**
### Summary

When a nested configuration error occurs, a more precise and identifiable description is required
When nested configuration conflicted occurs, it returns the desired flatterned error

Please see the test cases for details of the problem

### Checklist

- [x] The Pull Request has tests
- [x] A changelog file has been created under `changelog/unreleased/kong` or `skip-changelog` label added on PR if changelog is unnecessary. [README.md](https://github.com/Kong/gateway-changelog/blob/main/README.md)
- [ ] There is a user-facing docs PR against https://github.com/Kong/docs.konghq.com - PUT DOCS PR HERE

### Issue reference

Fix FTI-6351/FTI-6318

### CODE REVIEW

File: kong/db/errors.lua [hunk 1]
```diff
@@ -862,6 +863,9 @@ do
       --   }
       -- }
       --
+      if type(err_t[ref.field]) == "string" then
```
Comment (outsinre) [inline kong/db/errors.lua:866]: Seems the error message is always `string`?

File: kong/db/schema/others/declarative_config.lua [hunk 2]
```diff
@@ -400,8 +400,12 @@ local function populate_references(input, known_entities, by_id, by_key, expecte
       if key and key ~= ngx.null then
         local ok = add_to_by_key(by_key, entity_schema, item, entity, key)
         if not ok then
-          errs[entity] = errs[entity] or {}
-          errs[entity][i] = uniqueness_error_msg(entity, endpoint_key, key)
+          local errs_item_name = entity
+          if parent_entity and tostring(parent_idx) then
```
Comment (outsinre) [inline kong/db/schema/others/declarative_config.lua:404]: If `parent_idx` is `nil`, then `tostring(parent_idx)` is `true`.

File: kong/db/schema/others/declarative_config.lua [hunk 3]
```diff
@@ -400,8 +400,12 @@ local function populate_references(input, known_entities, by_id, by_key, expecte
...
+            errs_item_name = parent_entity.."|"..tostring(parent_idx).."|"..entity
+          end
+          errs[errs_item_name] = errs[errs_item_name] or {}
+          errs[errs_item_name][i] = uniqueness_error_msg(entity, endpoint_key, key)
```
Comment (outsinre) [inline kong/db/schema/others/declarative_config.lua:408]: The `errs_item_name` and `entity` on this line refer to the same thing?

File: changelog/unreleased/kong/fix-missing-flattened-erros.yml [hunk 4]
```diff
@@ -0,0 +1,4 @@
+message: |
+  Fixed an issue where flattened errors had missing info when fields were lost or conflicted in a nested entity.
```
Comment (outsinre) [inline changelog/unreleased/kong/fix-missing-flattened-erros.yml:2]:
```suggestion
  Fixed an issue where declarative flattened errors missed useful info when fields were absent or conflicted with each other in a nested entity.
```

File: spec/02-integration/04-admin_api/15-off_spec.lua [hunk 5]
```diff
@@ -536,6 +536,45 @@ describe("Admin API #off", function()
         assert.equals(snis.certificate.id, certificates.id)
       end)
 
+      it("snis in certificates without certificate", function()
```
Comment (outsinre) [inline spec/02-integration/04-admin_api/15-off_spec.lua:539]: Could we add a test with raw declarative config?

Or try `_format_version` with `3.0`?

File: kong/db/schema/others/declarative_config.lua (hunk 3 above)
Comment (lhanjian) [reply kong/db/schema/others/declarative_config.lua:408]: @outsinre no.

`entity` is its original name.
`errs_item_name` is key name of `errs`

File: kong/db/schema/others/declarative_config.lua (hunk 2 above)
Comment (outsinre) [reply kong/db/schema/others/declarative_config.lua:404]: The printed `nil` is a string, not the Lua `nil`.

File: spec/02-integration/04-admin_api/15-off_spec.lua [hunk 6]
```diff
@@ -536,6 +536,45 @@ describe("Admin API #off", function()
...
+                }
+              }
+            },
+          },
+          headers = {
+            ["Content-Type"] = "application/json"
+          }
+        })
+
+        local body = assert.response(res).has.status(400)
```
Comment (outsinre) [inline spec/02-integration/04-admin_api/15-off_spec.lua:573]: The `400` error is due to DB-less by default does not support this kind of format.

It expects the `snis` not nested within the `certificates`.

File: spec/02-integration/04-admin_api/15-off_spec.lua [hunk 7]
```diff
@@ -2323,6 +2362,76 @@ R6InCcH2Wh8wSeY5AuDXvu2tv9g/PW9wIJmPuKSHMA==
     }, post_config(input))
   end)
 
+  it("flatten_errors when conflicting inputs", function()
+    local conflicting_input = {
+      _format_version = "3.0",
+      consumers = {
```
Comment (flrgh) [inline spec/02-integration/04-admin_api/15-off_spec.lua:2368]: Hmm :thinking: I don't know if it's possible to properly flatten errors in this case, because the result we get from the declarative validator is ambiguous. Consider this input:

```lua
{
  _format_version = "3.0",
  consumers = {
    {
      username = "consumer-01",
      basicauth_credentials = {
        { username = "user-01", password = "pwd" },
        { username = "user-11", password = "pwd" },
        { username = "user-02", password = "pwd" },
        { username = "user-11", password = "pwd" },
      }
    },
    {
      username = "consumer-02",
      basicauth_credentials = {
        { username = "user-99", password = "pwd" },
      }
    },
    {
      username = "consumer-03",
      basicauth_credentials = {
        { username = "user-01", password = "pwd" },
        { username = "user-33", password = "pwd" },
        { username = "user-02", password = "pwd" },
        { username = "user-33", password = "pwd" },
      }
    }
  }
}
```

It has the following duplicates:

1. `consumers[1].basicauth_credentials[4]` (`user-11`)
2. `consumers[3].basicauth_credentials[1]` (`user-01`)
3. `consumers[3].basicauth_credentials[3]` (`user-02`)
4. `consumers[3].basicauth_credentials[4]` (`user-33`)

The error table we receive from `:parse_table()` looks like this:

```lua
{
  basicauth_credentials = {
    [1] = "uniqueness violation: 'basicauth_credentials' entity with username set to 'user-01' already declared",
    [3] = "uniqueness violation: 'basicauth_credentials' entity with username set to 'user-02' already declared",
    [4] = "uniqueness violation: 'basicauth_credentials' entity with username set to 'user-33' already declared",
  },
}
```

1. The declarative code actually overwrote the error in `consumers[1].basicauth_credentials[4]` with the error in `consumers[3].basicauth_credentials[4]` (presumably because they both happened at index `4`)
2. We've lost all context about which consumer the error corresponds to

In summary, I don't think this can be fixed in the error-flattening code alone. A fix within `kong.db.schema.others.declarative_config` must be made first.

**Comment by flrgh** [comment]
I actually fixed a very similar issue back in #12797. The problem is the fix was only applied for primary key errors, whereas the `basicauth_credentials.username` case is a unique field and not a primary key. I've applied the fix in #14017.

File: spec/02-integration/04-admin_api/15-off_spec.lua (hunk 7 above)
Comment (lhanjian) [reply spec/02-integration/04-admin_api/15-off_spec.lua:2368]: @flrgh 
I try to add your inputs to test case, it can get 4 results.

Is this what you need?

**Comment by outsinre** [comment]
There are two errors in this case.

1. The `basicauth_credentials` error caused by https://github.com/Kong/kong/blob/master/kong/db/schema/init.lua#L1504 but swallowed by https://github.com/Kong/kong/blob/358fff3869ce330880827a947db8bac3b5066215/kong/db/schema/others/declarative_config.lua#L792

   We'd better merge the two errors table.
2. The `snis` error caused by https://github.com/Kong/kong/blob/358fff3869ce330880827a947db8bac3b5066215/kong/db/schema/init.lua#L2134.
